// @dongowu/git-ai-cli/dist/commands/config.js
import inquirer from 'inquirer';
import chalk from 'chalk';
import { setConfig, getConfigPath } from '../utils/config.js';
import { PROVIDER_PRESETS } from '../types.js';
export async function runConfig() {
    console.log(chalk.cyan('\n🔧 Git AI Configuration\n'));
    const providerChoices = Object.entries(PROVIDER_PRESETS).map(([key, preset]) => ({
        name: preset.name,
        value: key,
    }));
    const { provider } = await inquirer.prompt([
        {
            type: 'list',
            name: 'provider',
            message: 'Select your AI provider:',
            choices: providerChoices,
        },
    ]);
    const preset = PROVIDER_PRESETS[provider];
    let baseUrl = preset.baseUrl;
    let apiKey = '';
    if (provider === 'custom') {
        const customAnswers = await inquirer.prompt([
            {
                type: 'input',
                name: 'baseUrl',
                message: 'Enter the API base URL:',
                validate: (input) => input.trim() ? true : 'Base URL is required',
            },
        ]);
        baseUrl = customAnswers.baseUrl;
    }
    if (preset.requiresKey) {
        const keyAnswer = await inquirer.prompt([
            {
                type: 'password',
                name: 'apiKey',
                message: 'Enter your API Key:',
                mask: '*',
                validate: (input) => input.trim() ? true : 'API Key is required',
            },
        ]);
        apiKey = keyAnswer.apiKey;
    }
    let model;
    let modelChoices = [];
    // Auto-detect Ollama models
    if (provider === 'ollama') {
        try {
            // Remove /v1 if present to access the tags endpoint
            const base = baseUrl.replace(/\/v1\/?$/, '');
            const response = await fetch(`${base}/api/tags`);
            if (response.ok) {
                const data = (await response.json());
                if (data.models && Array.isArray(data.models)) {
                    modelChoices = data.models.map((m) => m.name);
                }
            }
        }
        catch {
            // Fallback to manual input if connection fails
        }
    }
    if (modelChoices.length > 0) {
        const answer = await inquirer.prompt([
            {
                type: 'list',
                name: 'model',
                message: 'Select a local model:',
                choices: modelChoices,
            },
        ]);
        model = answer.model;
    }
    else {
        const answer = await inquirer.prompt([
            {
                type: 'input',
                name: 'model',
                message: 'Enter the model name:',
                default: preset.defaultModel,
            },
        ]);
        model = answer.model;
    }
    // Optional: separate model for Agent mode (tool calling)
    const modelLower = model.trim().toLowerCase();
    const suggestToolModel = provider === 'deepseek' && modelLower.includes('reasoner');
    const { useAgentModel } = await inquirer.prompt([
        {
            type: 'confirm',
            name: 'useAgentModel',
            message: suggestToolModel
                ? 'Agent mode needs tool calling. Use a tool-capable model for Agent mode?'
                : 'Use a separate model for Agent mode (tool calling)?',
            default: suggestToolModel,
        },
    ]);
    let agentModel = '';
    if (useAgentModel) {
        const agentAnswer = await inquirer.prompt([
            {
                type: 'input',
                name: 'agentModel',
                message: 'Enter the Agent mode model name:',
                default: suggestToolModel ? 'deepseek-chat' : model,
            },
        ]);
        agentModel = agentAnswer.agentModel.trim();
    }
    const { locale } = await inquirer.prompt([
        {
            type: 'list',
            name: 'locale',
            message: 'Select output language for commit messages:',
            choices: [
                { name: 'English', value: 'en' },
                { name: '中文', value: 'zh' },
            ],
        },
    ]);
    const { configurePrompt } = await inquirer.prompt([
        {
            type: 'confirm',
            name: 'configurePrompt',
            message: 'Would you like to configure a custom system prompt?',
            default: false,
        },
    ]);
    let customPrompt;
    if (configurePrompt) {
        const promptAnswer = await inquirer.prompt([
            {
                type: 'editor',
                name: 'customPrompt',
                message: 'Enter your custom system prompt:',
            },
        ]);
        customPrompt = promptAnswer.customPrompt.trim() || undefined;
    }
    const { enableFooter } = await inquirer.prompt([
        {
            type: 'confirm',
            name: 'enableFooter',
            message: 'Enable "Generated by git-ai" footer?',
            default: true,
        },
    ]);
    const config = {
        provider,
        apiKey,
        baseUrl,
        model,
        agentModel,
        locale,
        customPrompt,
        enableFooter,
    };
    setConfig(config);
    console.log(chalk.green('\n✅ Configuration saved successfully!'));
    console.log(chalk.gray(`   Config file: ${getConfigPath()}`));
    console.log(chalk.cyan('\n💡 You can now use `git-ai` to generate commit messages.\n'));
}
